Incrementally transfer chunks per token to improve handover

rfratto opened the pull request:
Design document: https://docs.google.com/document/d/1y2TdfEQ9ZKh6CpBVB4o6BYjCr-plNRL9jGD6fJ9bMW0/edit#

This PR introduces two _incremental chunk transfer_ process utilized by the lifecycler to reduce spillover and enable dynamic scaling of ingesters. The incremental transfer process takes precedence over the old handover mechanism. 

To migrate a cluster to use incremental transfers, two rollouts must be done:

1. Rollout ingesters with `-ingester.leave-incremental-transfer=true`
2. Rollout ingesters with `-ingester.join-incremental-transfer=true`

I recognize this is a large PR and I have attempted (to the best of my ability) to split it into smaller, independent commits. It's not perfect, but hopefully the commits I have make it easier to review. 

Fixes #1277.

/cc @gouthamve @pstibrany @tomwilkie

pstibrany commented:
There are multiple test failures and reported race conditions. Would be nice to fix those.

rfratto commented on pkg/ring/ring.go:
```diff
@@ -32,9 +32,10 @@ type ReadRing interface {
+	// ingeters and tokens are slices to be overwritten for the return value
+	// to avoid memory allocations; can be nil.
+	Get(key uint32, op Operation, ingesters []IngesterDesc,
+		tokens []TokenDesc) (ReplicationSet, error)
```
> Why the line break? it doesn't look very nice.

I've been adding line breaks >80 columns in most cases so it fits on my tiny terminal window without scrolling horizontally. I can merge this into one line if it's that much of an issue. 

> What is 'tokens' parameter used for? What is expected value after calling this method? Documentation missing.

It says it right at the top: 

> // ingeters and tokens are slices to be overwritten for the return value
> // to avoid memory allocations; can be nil.

pracucci replied:
> I've been adding line breaks >80 columns

My personal perspective: it's 2019, and the screen resolution goes well beyond 80 characters and restricting it to 80 is worse than not restricting at all. I personally see this kind of line breaks harder to read.

rfratto replied:
I agree 80 as a hard limit is a bit low, but I do think in general long lines should be broken up. I don't want to try suggesting what a soft limit should be, though. For now, I'm going to change this into one line.

In the case of other functions whose signatures are getting unwieldy, would either of you find one parameter per line easier to read?

```go
Get(
  key uint32,
  op Operation,
  ingesters []IngesterDesc,
  tokens []TokenDesc,
) (ReplicationSet, error) 
```

pstibrany replied:
(moved under this thread, my email comment was previously top-level)

First of all, I want to apologize for language used in my previous comment. I wanted to say the same thing as Marco did, but he did it in a much nicer way.

My IDE shows a column line at column 120, and I rarely see code that far (although there are lines crossing it in Cortex). If method signature got that long, I would first reconsider if it has to be like that. If so, I don’t have strong preference between the two styles [although slightly I prefer to use horizontal space, more than vertical], as long as I can see all arguments at onces, without scrolling. I agree with you that horizontal scrolling is annoying.

pracucci replied:
> In the case of other functions whose signatures are getting unwieldy, would either of you find one parameter per line easier to read?

The oneline version of your example is 100 columns wide. I still don't see the real need to split it.

```
Get(key uint32, op Operation, ingesters []IngesterDesc, tokens []TokenDesc) (ReplicationSet, error)
```

From my **personal** perspective: either this max columns is enforced somehow (and tested with the linter), or it's not a thing at all, otherwise we'll end up having different contributors adopting different limits. In the long term, consistency pays off more than perfection. Given the relatively long history of Cortex and the large code base, I would argue this is not a real issue worth to invest our effort, as far as we're OK relaxing the columns limit and we all write code within a reasonable width (ie. the current state of the code base looks good to me).

rfratto replied:
I was using the `Get` function as my example just to demonstrate another way one could shorten the length of a function signature; I was not trying to present it as it being too long, sorry. 

When I split lines at roughly 80 chars, I'm doing it as part of my style; i.e., I find code easier to consume the less I have to move my eyes horizontally (and my terminal window usually isn't very wide either). I don't think we should enforce a particular opinion on the ideal line length just as we shouldn't enforce when one should remove or add a blank line for vertical spacing. It's clearly going to be subjective when a line is "too long," and I agree that this conversation isn't worth our time right now.

My only concern here is doing line breaks in a way that doesn't make the code overall harder for readers to consume. You'll find that there are many places within this PR where I've used line breaks to split up long lines, particularly when calling functions. If you come across these and find them hard to read, please point them out, and I will find another solution to keep the lines short. 

For now, I will improve the documentation for `Get` and merge its function signature into one line.

pstibrany commented:
First of all, I want to apologize for language used in my previous comment. I wanted to say the same thing as Marco did, but he did it in a much nicer way.

My IDE shows a column line at column 120, and I rarely see code that far (although there are lines crossing it in Cortex). If method signature got that long, I would first reconsider if it has to be like that. If so, I don’t have strong preference between the two styles, as long as I can see all arguments at onces, without scrolling. I agree with you that scrolling is annoying.


> On 30 Oct 2019, at 21.21, Robert Fratto <notifications@github.com> wrote:
> 
> ﻿
> @rfratto commented on this pull request.
> 
> In pkg/ring/ring.go:
> 
> > @@ -32,9 +32,10 @@ type ReadRing interface {
>  	prometheus.Collector
>  
>  	// Get returns n (or more) ingesters which form the replicas for the given key.
> -	// buf is a slice to be overwritten for the return value
> -	// to avoid memory allocation; can be nil.
> -	Get(key uint32, op Operation, buf []IngesterDesc) (ReplicationSet, error)
> +	// ingeters and tokens are slices to be overwritten for the return value
> +	// to avoid memory allocations; can be nil.
> +	Get(key uint32, op Operation, ingesters []IngesterDesc,
> +		tokens []TokenDesc) (ReplicationSet, error)
> I agree 80 as a hard limit is a bit low, but I do think in general long lines should be broken up. I don't want to try suggesting what a soft limit should be, though. For now, I'm going to change this into one line.
> 
> In the case of other functions whose signatures are getting unwieldy, would either of you find one parameter per line easier to read?
> 
> Get(
>   key uint32,
>   op Operation,
>   ingesters []IngesterDesc,
>   tokens []TokenDesc,
> ) (ReplicationSet, error) 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

rfratto commented on pkg/ring/model.go:
```diff
@@ -75,14 +151,26 @@ func (d *Desc) RemoveIngester(id string) {
+				result = append(result, stok)
+			}
+
 			fromDesc.Tokens = nil
```
Good catch, thanks!

rfratto commented on pkg/ring/model.go:
```diff
@@ -147,14 +263,302 @@ func (d *Desc) Ready(heartbeatTimeout time.Duration) error {
+		d.Ingesters = make(map[string]IngesterDesc)
+	}
+
+	ing, ok := d.Ingesters[id]
```
Good point, I've taken out the check.

pstibrany commented on pkg/ring/incremental_transfer.go:
```diff
@@ -0,0 +1,386 @@
+				}
+
+				if err != nil {
+					level.Error(util.Logger).Log(
```
Please don't split these log messages to 6 lines (here, and elsewhere in this method). It's just a single log message, but this way of splitting it adds lot of visual clutter and it makes it difficult to quickly see the flow of the code.

rfratto replied:
I need to think about this one. I personally don't find it hard to read and it scans very similarly to how structs are instantiated. I get the argument that it's just a log line, but `level.Error(util.Logger).Log("msg", "failed to send chunks", "target_addr", addr, "ranges", PrintableRanges(ranges), "err", err)` would also bring attention to it and disrupt the flow of reading just by its sheer length. 

@gouthamve Is there a style standard for Cortex w/r/t line breaks?

pstibrany replied:
This is what I talk about ... 15 lines of visual clutter for what’s effectively 3 log messages. <img width="1194" src="https://user-images.githubusercontent.com/895919/67987599-26227b80-fc2e-11e9-899c-55f381cbbd81.png">

pstibrany replied:
My guess would be: https://github.com/golang/go/wiki/CodeReviewComments

rfratto commented on pkg/ring/lifecycler.go:
```diff
@@ -375,6 +559,69 @@ heartbeatLoop:
+// checkCleanRing returns true when the ring has no JOINING
+// or LEAVING ingesters. "clean" implies that it is safe for a
+// new node to join.
+func (i *Lifecycler) checkCleanRing(ctx context.Context) (bool, error) {
```
The best case scenario is that both ingesters split two distinct sets of ranges, although this is unlikely. The worst case scenario is that both ingesters will likely request one or more of the same ranges from the same target ingester, leading to an extra unintended replica.

rfratto commented on pkg/ring/lifecycler.go:
```diff
@@ -307,11 +462,39 @@ func (i *Lifecycler) loop() {
 	defer heartbeatTicker.Stop()
 
+	if i.cfg.JoinIncrementalTransfer {
+		autoJoinTimer.Stop()
```
Good point, thanks.

rfratto commented on pkg/ring/model.go:
```diff
@@ -147,14 +264,298 @@ func (d *Desc) Ready(heartbeatTimeout time.Duration) error {
+	// Start is the starting token to search from.
+	Start StatefulToken
+
+	// Neighbor is the number neighbor to search for.
```
How about Offset?

rfratto commented on pkg/ring/model.go:
```diff
@@ -34,25 +48,87 @@ func NewDesc() *Desc {
+		found := false
+		for i, t := range d.Tokens {
+			if t.Token == token.Token {
+				d.Tokens[i].State = token.State
```
I'm not sure I understand when that could happen. Could you explain?

pstibrany replied:
Well, that was my question to you :-) It probably cannot, but I need to check it to be sure.

rfratto replied:
I think the answer is technically yes, but likely not a concern. It _is_ possible that an ingester could generate the same token as another ingester before that token is inserted into the ring, but given that we attempt to ensure only one ingester can join at a time, I don't think we'll run into this.

rfratto commented on pkg/util/test/poll.go:
```diff
@@ -3,12 +3,13 @@ package test
 
 // Poll repeatedly evaluates condition until we either timeout, or it succeeds.
-func Poll(t *testing.T, d time.Duration, want interface{}, have func() interface{}) {
+func Poll(t require.TestingT, d time.Duration, want interface{}, have func() interface{}) {
```
Ah, I didn't know `testing.TB` existed, thanks!

rfratto commented on pkg/ring/model.go:
```diff
@@ -147,14 +264,298 @@ func (d *Desc) Ready(heartbeatTimeout time.Duration) error {
+			idx = len(d.Tokens) - 1
+		} else {
+			idx %= len(d.Tokens)
+		}
```
Unfortunately, I find that difficult to understand. I know this is a lengthy block but I think it's easy to see what it's doing and validate that it's doing it correctly.

pstibrany replied:
Fair enough, it was half-joke :)

rfratto commented on pkg/ring/token_checker.go:
```diff
@@ -0,0 +1,253 @@
+func (cfg *TokenCheckerConfig) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.BoolVar(&cfg.CheckOnCreate, prefix+"token-checker.check-on-create", false, "Check that newly created streams fall within expected token ranges")
+	f.BoolVar(&cfg.CheckOnCreate, prefix+"token-checker.check-on-append", false, "Check that existing streams appended to fall within expected token ranges")
+	f.BoolVar(&cfg.CheckOnCreate, prefix+"token-checker.check-on-transfer", false, "Check that streams transferred in using the transfer mechanism fall within expected token ranges")
```
Oops, just noticed I'm binding the same variable for all flags. Fix incoming soon.

rfratto commented on pkg/ring/token_checker.go:
```diff
@@ -0,0 +1,253 @@
+	// Used by ingester when set to true.
+	CheckOnCreate   bool `yaml:"check_on_create,omitempty"`
+	CheckOnAppend   bool `yaml:"check_on_append,omitempty"`
+	CheckOnTransfer bool `yaml:"check_on_transfer,omitempty"`
```
That's a fair point. I think I mostly agree, although it's somewhat nice to group them logically since they are all related. 

I'll move them to the ingester.

rfratto commented on pkg/ring/token_checker.go:
```diff
@@ -0,0 +1,253 @@
+					return
+				}
+
+				level.Debug(util.Logger).Log(
```
I may just take this out entirely. Even if it only prints when the expected range changes, it's still going to print numtokens^2 times. I don't think it adds a lot of value, anyway.

rfratto commented on pkg/ring/token_checker.go:
```diff
@@ -0,0 +1,253 @@
+
+	for _, tok := range tokens {
+		for replica := 0; replica < tc.lifecycler.cfg.RingConfig.ReplicationFactor; replica++ {
+			endRanges, err := tc.ring.Predecessors(NeighborOptions{
```
Yes, I agree that it's really expensive. I've tried many attempts to implement something less computationally complex for `Predecessors`, but couldn't in a reasonable amount of time. 

The primary issue is that when we search forwards through the ring, we only accept the _first_ token from each unique ingester. If we search backwards, we're only accepting the _last_ token from each unique ingester. I wasn't able to come up with an algorithm that searches backwards and also accounts for this difference. 

If you have any ideas on how else we can find all of the replica sets a token is in, please let me know, I'd like to make this better.

rfratto replied:
It may help if we only run `updateExpectedRanges` when the set of tokens change. Right now, it's going to run with every heartbeat.

rfratto replied:
I've come up with a way to improve `Predecessors` to be around `O(offset)`, in the average case depending on how the ring is arranged, and will make this operation a little less expensive, roughly `O(#ingesters * #tokens_per_ingester * replicationFactor^2)`. I'll implement it tomorrow and tag you so you can take a look.

pstibrany replied:
I was just thinking about the same issue... perhaps some kind of ring cleanup at the beginning would help? It would remove tokens for unhealthy ingesters, merge adjacent tokens for same ingesters [not 100% sure this is safe, but it seems like it should be]. That would simplify the data, and could possibly make code navigating it simpler as well?

rfratto commented on pkg/ingester/incremental_transfer.go:
```diff
@@ -0,0 +1,363 @@
+	defer i.blockedTokenMtx.Unlock()
+
+	for _, rg := range req.Ranges {
+		for _, r := range i.blockedTokens {
```
It needs to start the next iteration of outer loop. Missed this when I changed all the operations from using one token range to multiple, thanks

rfratto commented on pkg/ring/lifecycler.go:
```diff
@@ -375,6 +559,69 @@ heartbeatLoop:
+// waitCleanRing incrementally reads from the KV store and waits
+// until there are no JOINING or LEAVING ingesters.
+func (i *Lifecycler) waitCleanRing(ctx context.Context) error {
+	// Sleep for a random period up to 2s. Used to stagger
```
Could you explain? Would we need to set a minimum and maximum stagger? As it is now, it sleeps within the range 0-2s.

pstibrany replied:
With gossiping ring, there is a delay between ingester setting its state to JOINING/LEAVING and other ingesters seeing that -- it can easily be 5-10 seconds or even more, depending on size of the cluster and gossiping parameters.

pstibrany replied:
Aaah, I've misunderstood what this timeout does. Sorry for confusion.

rfratto replied:
I think you make a good point though, we also want to give new ingesters enough time to see the current view of the ring. I think adding a min and max jitter would help with that.

rfratto commented on pkg/ring/token_checker.go:
```diff
@@ -0,0 +1,253 @@
+func (tc *TokenChecker) watchRing(ctx context.Context) {
+	numIngesters := 0
+
+	tc.lifecycler.KVStore.WatchKey(ctx, ConsulKey, func(v interface{}) bool {
```
That should work. The TokenChecker can deal with only getting updates every heartbeat provided operators understand that temporary spikes in the metrics are expected. 

I'm going to work on rebasing this commit on top of master before addressing your comments; there's a lot of stuff I need to bring in.

rfratto commented on pkg/ring/lifecycler.go:
```diff
@@ -103,16 +124,64 @@ type FlushTransferer interface {
 	TransferOut(ctx context.Context) error
 }
 
+// TokenRange represents a range of tokens, starting inclusively
```
I don't think so - that would be a circular dependency between ring and the ingester, and we also don't want Loki to have to depend on the Cortex ingester code.

pstibrany replied:
Good point, I didn't realize that.

rfratto replied:
I'm going to do this the other way around: I'll move TokenRange to `ring.proto` and import it from `cortex.proto`.

rfratto commented:
I've addressed most of the review feedback so far. I want to rebase against latest and fix the merge conflicts before I continue addressing feedback. This may take a little bit of time; both the TSDB blocks and the gossip are going to change bits and pieces about the current implementation.

pstibrany commented on pkg/ring/token_checker.go:
```diff
@@ -0,0 +1,253 @@
+	// found. If nil, is a no-op.
+	UnexpectedTokenHandler func(tokens []uint32)
+
+	mut        sync.Mutex
```
What I actually meant was to group it together with `ring` and `expectedRanges`, because it protects those two fields.

```
	mut            sync.Mutex
	ring           *Desc
	expectedRanges []TokenRange
```

pstibrany replied:
This code is still little confused about mutex and whether to use it or not for context and cancel function. Shutdown uses mutex for calling cancel(), but Start doesn't use mutex for setting it.

(I liked previous design better, when NewTokenChecker created context and started go-routines as well... there was no need to split that, except for setting UnexpectedTokenHandler, but that could have been passed to NewTokenChecker method instead. Then mutex wouldn't be necessary for context/cancel).

pstibrany commented on pkg/ring/incremental_transfer.go:
```diff
@@ -0,0 +1,391 @@
+					)
+				}
+
+				wg.Done()
```
It's a good practice to call `wg.Done()` via `defer` at the beginning of the function (multiple places in this PR)

pstibrany commented on pkg/ring/incremental_transfer.go:
```diff
@@ -0,0 +1,391 @@
+					)
+				}
+
+				wg.Done()
```
Why split this logic into  multiple goroutines? One doing the `BlockRanges`, other one doing `SendChunkRanges` and `UnblockRanges`?

rfratto replied:
We defer unblocking ranges after a delay to make sure we give enough time for distributors to see the new ACTIVE token in the ring, which is why it's done in a separate goroutine.

pstibrany commented on pkg/ring/incremental_transfer.go:
```diff
@@ -0,0 +1,391 @@
+// joinIncrementalTransfer will attempt to incrementally obtain chunks from
+// neighboring lifecyclers that contain data for token ranges they will
+// no longer receive writes for.
+func (i *Lifecycler) joinIncrementalTransfer(ctx context.Context) error {
```
This method is too long, and would really benefit from some refactoring. For example handling of individual token can be extracted into a separate method? Same goes for `joinIncrementalTransfer`.

rfratto replied:
Sure, I'll do that.

rfratto commented on pkg/ring/model.go:
```diff
@@ -147,14 +272,331 @@ func (d *Desc) Ready(heartbeatTimeout time.Duration) error {
+		ing := d.Ingesters[predecessor.Ingester]
+
+		// Stop if this is a new ingester and we already have seen enough unique ingesters.
+		if _, ok := distinct[predecessor.Ingester]; !ok && len(distinct) == opts.Offset+1 {
```
No, we want to collect all the tokens from ingesters that _may_ be a predecessor, and stop only after we see a token from an ingester that _can't_ be a predecessor. That's defined here as `opts.Offset+1` ingesters (i.e.,the `opts.Offset+2` ingester can never be a predecessor).

rfratto commented on pkg/ring/model_test.go:
```diff
@@ -0,0 +1,230 @@
+		expect []uint32
+	}{
+		// Simple cases
+		{"A B C", 2, 0, Read, []uint32{3}},
```
I agree this is confusing. The first number is the token's index in the ring, while the expected number is the token's value (i.e., `index + 1`). I'll change all the tests to be consistent and use the token values. I struggle reading this myself, and I wrote it.

rfratto commented:
> Main issue I have is with the requirement that we need to move all the replicated data around, not just the primary data for token. This is further complicated by adjacent tokens belonging to the same ingester, dealing with unhealhty ingesters and token states.

Adjacent tokens belonging to the same ingester, I find, is the easier half of the problem. The main complexities with dealing with the ring are when tokens belonging to the same ingester are near, but not next to, each other. For example, when dealing with a ring `A1 B A2 C`, we have to keep the tokens unmerged to know which subset of ranges `A` will be handling. Dealing with tokens that are next to each other is handled implicitly through dealing with tokens that are near each other. 

I don't think we can loosen the requirement of not moving all replicated data around. If we didn't do this, every time an ingester leaves, we would lose `1/replicationFactor` of our data. If you did a complete rollout, you would lose all replicas outside of the original owner. This would subtly break a lot of things, including querying, which use the quorum to return correct results.

pstibrany commented:
> I don't think we can loosen the requirement of not moving all replicated data around. If we didn't do this, every time an ingester leaves, we would lose `1/replicationFactor` of our data. If you did a complete rollout, you would lose all replicas outside of the original owner. This would subtly break a lot of things, including querying, which use the quorum to return correct results.

Can you elaborate on how we would lose data? In the complete rollout scenario, we can use the same mechanism as we do today. Perhaps the solution could be to warn admin about not leaving too many ingesters at once?

rfratto commented:
> Can you elaborate on how we would lose data? In the complete rollout scenario, we can use the same mechanism as we do today. Perhaps the solution could be to warn admin about not leaving too many ingesters at once?

We would lose data because each ingester holds data for `replicationFactor` total ingesters , including itself. My fallback is to flush anything that didn't get transferred, but we generally have queriers configured to only query the store for stuff that isn't in memory anymore. That means there will be some period where an ingester gets a query for some data that it should be a replica of, but it never received that data during the transfer.

pstibrany commented:
My understanding is that queriers ask all ingesters, so any ingester with data will reply.

> On 5 Nov 2019, at 14.09, Robert Fratto <notifications@github.com> wrote:
> 
> ﻿
> Can you elaborate on how we would lose data? In the complete rollout scenario, we can use the same mechanism as we do today. Perhaps the solution could be to warn admin about not leaving too many ingesters at once?
> 
> We would lose data because each ingester holds data for replicationFactor total ingesters , including itself. My fallback is to flush anything that didn't get transferred, but we generally have queriers configured to only query the store for stuff that isn't in memory anymore. That means there will be some period where an ingester gets a query for some data that it should be a replica of, but it never received that data during the transfer.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

rfratto commented:
> My understanding is that queriers ask all ingesters, so any ingester with data will reply.

I may be slightly wrong, but I think queriers ask all ingesters and stop once it receives responses from a quorum number of those ingesters. If none of the quorum had any data, then the query results would show no data. Again, unsure, but I believe this is how it works.

pstibrany commented:
You may by right, I haven’t checked the code, although why would it need quorum? Either it gets data from somewhere, or not. It cannot possibly get different data from different ingesters (perhaps incomplete?). I need to check the code.

> On 5 Nov 2019, at 20.01, Robert Fratto <notifications@github.com> wrote:
> 
> ﻿
> My understanding is that queriers ask all ingesters, so any ingester with data will reply.
> 
> I may be slightly wrong, but I think queriers ask all ingesters and stop once it receives responses from a quorum number of those ingesters. If none of the quorum had any data, then the query results would show no data. Again, unsure, but I believe this is how it works.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

rfratto commented on pkg/ring/model.go:
```diff
@@ -47,6 +48,58 @@ func NewDesc() *Desc {
 
+// AddToken adds the given token into the ring for a specific ingester.
+// If the token already exists for that ingester, it will be overwritten.
+func (d *Desc) AddToken(id string, token StatefulToken, normaliseTokens bool) {
```
Can this be merged with RefreshTokenState?

rfratto commented on pkg/ingester/transfer.go:
```diff
@@ -291,3 +400,396 @@ func (i *Ingester) findTargetIngester(ctx context.Context) (*ring.IngesterDesc,
+		})
+
+		blockedRanges.Inc()
+	}
```
This may be simpler if we just have blockedTokens be a map with the key being the range.

rfratto commented on pkg/ring/token_checker.go:
```diff
@@ -0,0 +1,234 @@
+func (tc *TokenChecker) CheckAllTokens() bool {
+	var invalid []uint32
+
+	toks := tc.lifecycler.incTransferer.StreamTokens()
```
You're right, although I think that's what we want - we want the total number of unexpected streams here even if they share tokens. I'll update the names `CheckallStreams`, `UnexpectedStreamsHandler` to reflect this better.

rfratto commented on pkg/ring/model.go:
```diff
@@ -442,3 +642,264 @@ func (d *Desc) RemoveTombstones(limit time.Time) {
+
+			distinct[successor.Ingester] = struct{}{}
+		}
+	}
```
I suspect both `Predecessors` and `Successors` would be slightly easier to read if they called upon a ring iterator that abstracted out handling token indices, health checks, and wrapping around the ring.

rfratto commented on pkg/ring/lifecycler.go:
```diff
@@ -631,6 +897,18 @@ func (i *Lifecycler) changeState(ctx context.Context, state IngesterState) error
+		i.setTokensState(state)
+	} else if !i.cfg.LeaveIncrementalTransfer && state == LEAVING {
+		i.setTokensState(state)
+	}
```
A better way of handling this is to conditionally enable whether or not the ingester is using stateful tokens.

rfratto replied:
I'm backpedaling on this. I tried implementing it but it's more complicated to follow than having this here.

rfratto commented on pkg/ring/lifecycler.go:
```diff
@@ -673,6 +967,21 @@ func (i *Lifecycler) processShutdown(ctx context.Context) {
+	}
+	desc.Tokens = migrateRing(desc)
+	return desc, nil
+}
```
`updateConsul` would have already just been called any time this function is invoked. We can save the most recently `CAS`'d value and use that instead here. The stored value can be denormalized so the TokenChecker can also use it.

rfratto commented:
I need to figure out why the tests time out on CircleCI, but I've refactored a bit and addressed most of the PR comments. PTAL, @pstibrany

gouthamve commented:
Fixes #467

pstibrany commented on pkg/ring/model.go:
```diff
@@ -11,13 +12,44 @@ import (
 func (ts ByToken) Less(i, j int) bool { return ts[i].Token < ts[j].Token }
 
+// byUint32 is a sortable list of uint32
+type byUint32 []uint32
```
This type is not used anywhere, remove

pstibrany commented on pkg/ring/model.go:
```diff
@@ -168,28 +202,106 @@ func (d *Desc) Ready(now time.Time, heartbeatTimeout time.Duration) error {
+	if normalise {
+		ing.AddTokens(tokens)
+		d.Ingesters[id] = ing
+	} else {
```
Hopefully https://github.com/cortexproject/cortex/pull/1809 gets merged soon (reviews welcome! :-)), so that this code can be removed.

pstibrany commented on pkg/ring/replication_strategy.go:
```diff
@@ -2,51 +2,51 @@ package ring
+// replicationStrategy will check to see if there's enough tokens for
+// an operation to succeed; unhealthy tokens are filtered out using
+// IsHealthyState.
+func (r *Ring) replicationStrategy(tokens []TokenDesc, op Operation) ([]TokenDesc, int, error) {
```
I would love to remove this from `*Ring`, either to be standalone function, or move to `*Desc` instead. It only uses `r.cfg.ReplicationFactor` from `r`. [Optional]

pstibrany commented on pkg/ring/ring.proto:
```diff
@@ -8,28 +8,31 @@ option (gogoproto.marshaler_all) = true;
 message Desc {
-	map<string,IngesterDesc> ingesters = 1 [(gogoproto.nullable) = false];
-	repeated TokenDesc tokens = 2 [(gogoproto.nullable) = false];
+  map<string,IngesterDesc> ingesters = 1 [(gogoproto.nullable) = false];
```
Diff still shows tabs replaced with spaces in entire file? Please try to minimize such changes. 
<img width="1239" alt="Screenshot 2019-11-15 at 16 23 55" src="https://user-images.githubusercontent.com/895919/68954455-53d9f980-07c4-11ea-9faf-2ad9e90138af.png">

rfratto commented on pkg/ring/model_test.go:
```diff
@@ -175,3 +176,149 @@ func TestReady(t *testing.T) {
+
+		// only consider healthy tokens
+		{"A B C- D E", RangeOptions{Range: TokenRange{1, 5}, ID: "C"}, true},
+		{"A B C- D E", RangeOptions{Range: TokenRange{1, 5}, ID: "C"}, true},
```
`C-` is healthy for a read operation, but you're right about the other cases, I've removed them.

rfratto commented on pkg/ring/model.go:
```diff
@@ -442,3 +554,267 @@ func (d *Desc) RemoveTombstones(limit time.Time) {
+	}
+}
+
+// NeighborOptions holds options to configure how a ring will be searched for
```
Good point, it used to have a lot more fields before the refactor. It probably doesn't need to be around anymore.

pstibrany commented on pkg/ring/incremental_transfer.go:
```diff
@@ -0,0 +1,252 @@
+
+	r := i.getDenormalisedRing(ctx)
+
+	replicationFactor := i.cfg.RingConfig.ReplicationFactor
```
This is an optimization for cases when number of active ingesters is low. Let's put some comment here.

rfratto commented on pkg/ingester/client/cortex.proto:
```diff
@@ -20,11 +21,25 @@ service Ingester {
+  // CleanupTransfer is invoked by a joining ingester after it is
+  // safe to perform cleanup on anything initialized by a call to
+  // GetChunks.
+  rpc CleanupTransfer(CleanupTransferRequest) returns (CleanupTransferResponse) {};
```
This ends up just unblocking blocked ranges, but I'm hesitant to add any more context at the RPC level. I made it generic on purpose since Loki may do more than that. WDYT?

rfratto commented on pkg/ring/lifecycler.go:
```diff
@@ -631,9 +868,27 @@ func (i *Lifecycler) changeState(ctx context.Context, state IngesterState) error
+	return i.lastRing
+}
+
 func (i *Lifecycler) updateCounters(ringDesc *Desc) {
```
I'd like to keep `updateCounters` since it's used for caching the results for the number of healthy instances. I'll move updating the last saved ring into another function.

pstibrany replied:
It is slightly strange that `lastRing` is protected by `i.countersLock` though :)

pstibrany commented on pkg/ring/lifecycler.go:
```diff
@@ -246,18 +297,73 @@ func (i *Lifecycler) ChangeState(ctx context.Context, state IngesterState) error
+	i.tokens = append(i.tokens, token)
+}
+
+func (i *Lifecycler) removeToken(token uint32) {
```
```
func (i *Lifecycler) removeToken(token uint32) {
	i.stateMtx.Lock()
	defer i.stateMtx.Unlock()

	for idx, tok := range i.tokens {
		if tok == token {
			tokensOwned.WithLabelValues(i.RingName).Dec()
			i.tokens = append(i.tokens[:idx], i.tokens[idx+1:]...)
			return
		}
	}
}
```

pstibrany commented on pkg/ring/incremental_transfer.go:
```diff
@@ -0,0 +1,254 @@
+	}
+
+	if !success {
+		return fmt.Errorf("incremental transfer out incomplete")
```
Single failure to updateConsul will cause this error, but if the last updateConsul succeeded and everything else was fine, perhaps it can still be considered as a success? Can we reuse call to `checkRemainingTokens` here to check if all is fine?

pstibrany commented on pkg/ring/model.go:
```diff
@@ -442,3 +503,229 @@ func (d *Desc) RemoveTombstones(limit time.Time) {
+// Neighbor moves around the ring to find the nth neighboring ingester of start.
+// The healthy function determines which tokens are considered as neighbors.
+//
+// Only one token per ingester is considred as a neighbor. If excludeStart is
```
ah, I forgot: please don't use "negatives" in parameter and field names. it's considered a bad style, and leads to confusion. can we call this parameter: "startIngesterCanBeNeighbor"? it's descriptive and doesn't use negation... (something shorter than that would be nice though)

rfratto commented on pkg/ring/incremental_transfer.go:
```diff
@@ -0,0 +1,252 @@
+
+	r := i.getDenormalisedRing(ctx)
+
+	replicationFactor := i.cfg.RingConfig.ReplicationFactor
```
Will do. But for reference, it's not an optimization, but a requirement. If the number of ingesters in the ring is less than the RF, some calls to Successor will fail when we try getting the target.

pstibrany replied:
> If the number of ingesters in the ring is less than the RF, some calls to Successor will fail when we try getting the target.

Ah, makes sense. That explains it and should be in the comment.

rfratto commented on pkg/ring/incremental_transfer.go:
```diff
@@ -0,0 +1,252 @@
+	for _, token := range i.transitioningTokens {
+		// Mark our token to JOINING and add it to our copy of the ring.
+		token.State = JOINING
+		r.RefreshTokenState(i.ID, append(i.getTokens(), token), false)
```
Yes, we're only doing it so that `findTransferWorkload` detects the new token properly. Other ingesters don't need to know about the new token until we've accepted all data for the  new token.

rfratto commented on pkg/ring/model.go:
```diff
@@ -104,10 +105,11 @@ func (d *Desc) ClaimTokens(from, to string, normaliseTokens bool) []uint32 {
 			i++
 		}
 
+		sort.Sort(sortableUint32(result))
```
I think it was added from a suggestion you made in a previous comment that it may be needed 😄 I'll take it out.

pstibrany replied:
That wasn't me! :) ... in other words, sorry about that :-) ... but to be fair, this method has changed several times, and we have only removed with Goutham recently when merging gossiping code, which was after my first reviews here :-) (ok, enough excuses now)

rfratto commented on pkg/ring/lifecycler.go:
```diff
@@ -634,6 +864,19 @@ func (i *Lifecycler) changeState(ctx context.Context, state IngesterState) error
+		return nil
+	}
+
+	r := *i.lastRing
```
The result of this function gets modified by `joinIncrementalTransfer` and `leaveIncrementalTransfer` and read simultaneously in another goroutine (token checker). It'd be a race condition to not return a copy - I'll explicitly copy the whole structure.

pstibrany replied:
Then it would make more sense to me to create a copy where it's needed, and not here.

rfratto replied:
It's safer to return a copy here and not at the call site. We don't want to allow callers to modify our copy of the last ring. I would say it's common practice to return copies in get methods rather than trusting callers to do the right thing.

pstibrany replied:
That is a safe thing to do when writing APIs for public consumption, but not always needed for internal usage, when you control all callers. For example token_checker.go doesn't need a copy. But it's ok.

rfratto commented on pkg/ring/model.go:
```diff
@@ -442,3 +503,229 @@ func (d *Desc) RemoveTombstones(limit time.Time) {
+	return TokenDesc{}, fmt.Errorf("could not find neighbor #%d for token %d", offset, start)
+}
+
+// Predecessors finds all tokens of which the start is the offset-th neighbor to.
```
So here's what's happening with the error messages: 

1. We want to know the set of tokens where 518 is the 5th successor (based on `5:`) there. 
2. We work backwards from 518, starting at 505. 
3. We get the 5th successor for 505 and check if it's 518. 
4. There is no 5th successor for 505. There are only 5 ingesters in the ring total, so there will be no more than 4 successors. 

However, something seems fishy with offsets 0-4 returning empty slices. Offset 1 should return `[{295 D}]`. How are you generating the ring?

pstibrany replied:
This is the test code I use: https://gist.github.com/pstibrany/4131a178dfa03866901dc1faf4f38568

(I had to modify some methods to be callable like this, see full.patch in same gist)

rfratto replied:
Took me a while to catch it, but the output is correct here. Because token `505` is right before `518`, `505` will always be the successor chosen by all the previous tokens. This is actually a case we _do_ want to catch, because it means that `518` has nothing to transfer for the non-zero offsets. 

For zero offsets (i.e., `518` itself), we'd still want to transfer stuff in the `505`-`518` range (which E owns), so this makes sense too.

pstibrany replied:
Thanks for explanation, it makes sense to me now.

rfratto commented on pkg/ring/token_checker.go:
```diff
@@ -0,0 +1,194 @@
+
+// updateExpectedRanges goes through the ring and finds all expected ranges
+// given the current set of tokens in a Lifecycler.
+func (tc *TokenChecker) updateExpectedRanges(ring *Desc) {
```
Yes, it is OK. Even though C is in the ring, because it is unhealthy, A will be receiving all of C's writes from the distributor; A effectively owns C's ranges. When C is healthy again, this function will properly detect that and shorten the ranges for A.

rfratto commented on pkg/ingester/incremental_transfer.go:
```diff
@@ -0,0 +1,264 @@
+
+	// Block the ranges locally so we don't receive traffic for it anymore.
+	// Will be later unblocked by CleanupTransfer.
+	if err := i.BlockRanges(stream.Context(), req.Ranges); err != nil {
```
We don't call `UnblockRanges` at the end of `GetChunks` because we want writes to continue to be rejected until the joining ingester has successfully added its token into the ring. `CleanupTransfer` is used to protect against spillover. The automatic unblock is used as a failsafe in case the joining ingester fails or crashes: writes will continue to go to the blocked ranges and be rejected. We want them to succeed eventually, so we let them unblock themselves after a timeout.

pstibrany replied:
I don't like spawning of new goroutine for each BlockRanges, would prefer single goroutine that expires blocked ranges, if any. To make what you write work properly, should joining ingester keep extending those blockages until it is ready to unblock it all?

rfratto replied:
I'm not sure how I feel about a permanent goroutine that waits for blocked ranges and sees if they need to be removed after an expiration period. It would also mean we'd have to introduce a new `TimedRange` (or something). I'll put some thought into it. 

> To make what you write work properly, should joining ingester keep extending those blockages until it is ready to unblock it all?

Implying that this doesn't work properly? 🙂 I feel that it's up to operators to configure the timeout to be longer than a join takes. Adding in extensions to the blocked range timeout seems like it would be a lot of unnecessary logic.

pstibrany replied:
But how can operator guess how long it will take? What if he guesses wrongly? On the other hand, ingester knows when it's ready, and if it's not but still running, it can extend the timeout.

pstibrany replied:
> I'm not sure how I feel about a permanent goroutine that waits for blocked ranges and sees if they need to be removed after an expiration period.

It can be spawned only when needed and not running yet. If goroutine unblocks everything, it can die again. (ok, that's somewhat more complex than what I had in mind. I think single goroutine running all the time and checking timeouts would be just fine).

rfratto replied:
> But how can operator guess how long it will take? What if he guesses wrongly?

It should just be set to a large value, overshooting shouldn't be a practical concern. The default timeout should be covering it for most clusters, and if you're running a giant cluster, shutdown duration is a known metric.

pstibrany replied:
Or it can be one less thing to worry about with little extra code? ;-) But I see your point. Basically you would just use large enough timeout to be sure it's ok.

rfratto commented on pkg/ring/incremental_transfer.go:
```diff
@@ -0,0 +1,254 @@
+		}
+
+		// Add the new token into the copy of the ring.
+		r.AddTokens(i.ID, append(i.getTokens(), token), false)
```
Because `r` is a copy of the ring that has been migrated and is always denormalized.

pstibrany replied:
OK, that makes sense. However, this code relies on a denormalization that we're trying to eradicate from the codebase, and only keep in migrateRing. Not sure how to resolve it yet.

rfratto commented on pkg/ring/incremental_transfer.go:
```diff
@@ -0,0 +1,254 @@
+			err := i.incTransferer.SendChunkRanges(ctx, ranges, addr)
+			if err != nil {
+				level.Error(util.Logger).Log("msg", "failed to send chunks", "target_addr", addr, "ranges", PrintableRanges(ranges), "err", err)
+				success = false
```
Is this true? `success` is only used for writing across the goroutines. Isn't `atomic.Bool` only needed if you want to synchronize both reads and writes?

pstibrany replied:
Read happens on different goroutine.

rfratto replied:
But the read only happens after the other writing goroutines are finished.

pstibrany replied:
```
func TestWorkloadConcurrency(t *testing.T) {
	tw := transferWorkload{
		"ing1": []TokenRange{{10, 20}, {30, 40}},
		"ing2": []TokenRange{{50, 60}, {70, 80}},
	}

	success := false
	tw.Do(func(addr string, ranges []TokenRange) {
		success = true
	})

	fmt.Println(success)
}
```

This fails with go test -race.

rfratto commented on pkg/ring/incremental_transfer.go:
```diff
@@ -0,0 +1,254 @@
+	}
+
+	go func() {
+		if i.cfg.TransferFinishDelay != time.Duration(0) {
```
The delay allows for the distributors to pick up the new tokens added by the joining ingester before we unblock the ranges, helping prevent spillover. We want to ensure that the old ingesters do not receive any more writes for tokens we are now responsible for.

rfratto commented on pkg/ingester/transfer.go:
```diff
@@ -52,57 +59,169 @@ func init() {
+	Send(*client.TimeSeriesChunk) error
+}
+
+type pushChunksOptions struct {
```
I'd like to push back on this one a bit. Four parameters is nearing too many, especially when one is optional. I'd like to leave the options struct around for this one.

pstibrany replied:
I think either both acceptChunks and pushChunks should use options struct, or neither should. In my opinion they don’t reach the imaginary parameters limit yet, and optional func is just `nil`.

rfratto commented on pkg/ingester/transfer.go:
```diff
@@ -52,57 +59,169 @@ func init() {
+	for userID, state := range opts.States {
+		curCh = state.fpToSeries.iter()
+		for pair := range curCh {
+			if opts.ExcludeFunc != nil && opts.ExcludeFunc(&pair) {
```
Better safe than sorry, I'll lock it.

rfratto commented on pkg/ingester/incremental_transfer.go:
```diff
@@ -0,0 +1,264 @@
+		return err
+	}
+
+	userStatesCopy := i.userStates.cp()
```
This one crosses into a lot of code I haven't looked at before. We use the fingerprint mapper to protect fingerprints, and the userStatesMtx when we're not modifying the copy. 

@jtlisi any ideas on if we need extra synchronization here?

pstibrany replied:
I think locking fingerprints when using shared states should be enough, but let’s see what others have to say.

rfratto commented on pkg/ingester/client/cortex.proto:
```diff
@@ -20,11 +21,25 @@ service Ingester {
+
+  // SendChunks is invoked by a leaving ingester (client), streaming a subset
+  // of its chunks directly to an existing ingester.
+  rpc SendChunks(stream TimeSeriesChunk) returns (SendChunksResponse) {};
```
I'm going to use Transfer for both and  add `Subset` or something similar. I'm afraid of reducing the functionality of RPC calls because handoff still needs to work from ingesters that aren't aware of the new incremental transfer functionality.

rfratto commented on pkg/ingester/client/cortex.proto:
```diff
@@ -125,6 +141,40 @@ message TimeSeries {
+message BlockRangeResponse {
+}
+
+message UnblockRangeResponse {
```
`google.protobuf.Empty` makes me nervous when we're using `gogo`, and I couldn't find `Empty` in our vendor of gogo. I'm going to leave this as-is for now.

pstibrany replied:
ok

rfratto commented on pkg/ingester/incremental_transfer.go:
```diff
@@ -0,0 +1,264 @@
+
+			if head := pair.series.head(); head != nil && !head.flushed {
+				ret = append(ret, pair.series.token)
+			}
```
Good catch.

pstibrany commented on pkg/ring/model.go:
```diff
@@ -168,6 +191,39 @@ func (d *Desc) Ready(now time.Time, heartbeatTimeout time.Duration) error {
 
+// SetIngesterTokens updates the tokens in the ring with the provided tokens. SetIngesterTokens
+// will remove denormalised tokens for id that are no longer in the list provided by tokens.
+func (d *Desc) SetIngesterTokens(id string, tokens []uint32, normalise bool) {
```
Can we move this closer to `AddIngester` method, to keep them together?

pstibrany commented on pkg/ring/model.go:
```diff
@@ -168,6 +191,39 @@ func (d *Desc) Ready(now time.Time, heartbeatTimeout time.Duration) error {
 	return nil
 }
 
+// SetIngesterTokens updates the tokens in the ring with the provided tokens. SetIngesterTokens
```
So I have an idea ... all of these new methods that you're introducing to ring.Desc -- I understand why you need them but I also don't like that it bloats the ring code, structure that is stored in KV store, so much.

The fact that there is still `ring.Desc.Tokens` field is just a legacy, and we aim to get rid of it, as we don't want to store it in the KV. But obviously code working with the ring in memory operates on that.

I think we should split ring as stored in KV and ring as used in memory (basically `[]TokenDesc`) and have them as two separate types. Many of your new methods would go to the new type then, and we would be free to get rid of `ring.Desc.Tokens` eventually.

pstibrany commented on pkg/ring/incremental_transfer.go:
```diff
@@ -0,0 +1,259 @@
+	go func() {
+		// Delay unblocking ranges to allow the distributors enough time to notice the
+		// new token. If ranges are unblocked too soon, spillover happens.
+		if i.cfg.TransferFinishDelay != time.Duration(0) {
```
Should be `>` not `!=`.

pstibrany commented on pkg/ring/token_checker.go:
```diff
@@ -0,0 +1,195 @@
+// in the ring change over time.
+type TokenChecker struct {
+	cfg        TokenCheckerConfig
+	rcfg       Config
```
`ringCfg`

pstibrany commented:
Great, thanks for latest batch of changes. Would it also be possible to
- make migrateRing to return `TokenNavigator`
- not store new value to d.Tokens in places like `(*Lifecycler).updateLastRing` and `(*Ring).loop` (only two places), and use new TokenNavigator only?

I'm still concerned about using calling `SetIngesterTokens` with `normalise = false` parameter, and would like to get rid of them. Ideally, such method would only be done on `TokenNavigator` instances. Goal is to remove Desc.Tokens completely, adding new places where we rely on it is therefore not good.

rfratto commented:
I can add an equivalent of `SetIngesterTokens` into the `TokenNavigator`, but we still need the function to exist in `ring.Desc` since the ingesters need to incrementally add their tokens over time.

I'll work on using `TokenNavigator` in more places, the challenge right now is passing around both the navigator and the ring desc, as we need the ring desc for the addresses of the ingesters.

pstibrany commented:
> I can add an equivalent of `SetIngesterTokens` into the `TokenNavigator`, but we still need the function to exist in `ring.Desc` since the ingesters need to incrementally add their tokens over time.

That's fine, but in `ring.Desc`, it would respect normalisation flag. It doesn't need to (and cannot) do that in `TokenNavigator`.

pstibrany commented on pkg/ring/token_navigator.go:
```diff
@@ -0,0 +1,269 @@
+
+// NewTokenNavigator creates a new TokenNavigator given a r, creating a
+// copy of the token list.
+func NewTokenNavigator(r *Desc) TokenNavigator {
```
nitpick: I would move this to *Desc type instead (`ToTokenNavigator` or `GetDenormalisedTokens`) – to avoid referencing *Desc from this file, and also to avoid passing *Desc as parameter, if it can have this as method directly. But it's a minor detail.

pstibrany commented on pkg/ring/token_navigator.go:
```diff
@@ -0,0 +1,269 @@
+	}
+
+	sort.Sort(ByToken(tokens))
+	return TokenNavigator(tokens)
```
this cast is redundant

pstibrany commented on pkg/ring/token_navigator.go:
```diff
@@ -0,0 +1,269 @@
+// TokenNavigator provides utility methods for traversing the ring.
+type TokenNavigator []TokenDesc
+
+// NewTokenNavigator creates a new TokenNavigator given a r, creating a
```
`given a ring`? (But by moving this method to *Desc, this can be reworded)

pstibrany commented on pkg/ring/model.go:
```diff
@@ -442,3 +523,17 @@ func (d *Desc) RemoveTombstones(limit time.Time) {
+
+// HealthCheckFunc is a function that validates whether a given token in the
+// ring is healthy and can be used.
+type HealthCheckFunc func(TokenDesc) bool
```
This should be moved to place where it is used (=consumed) - i.e. token_navigator.go.

pstibrany commented on pkg/ring/lifecycler.go:
```diff
@@ -634,6 +858,18 @@ func (i *Lifecycler) changeState(ctx context.Context, state IngesterState) error
+		return nil
+	}
+
+	return i.lastRing.Clone()
```
Do we still need to have the Clone method? I believe we no longer modify the returned instance?

owen-d commented on pkg/ingester/incremental_transfer.go:
```diff
@@ -0,0 +1,260 @@
+	}
+
+	// Target ingester may not have had any streams to send.
+	if seriesReceived == 0 {
```
This `if seriesReceived == 0` check seems unnecssary as you're returning nil regardless.

owen-d commented on pkg/ingester/incremental_transfer.go:
```diff
@@ -0,0 +1,260 @@
+	// when ranges should be unblocked: we should continue to reject writes for as long as we may
+	// receive them. When the joining token has been completely inserted into the ring, it will
+	// be safe to remove the blocks.
+	i.BlockRanges(req.Ranges)
```
shouldn't the block occur before copying user states in the unlikely case there's a new series appended during `i.userStates.cp()`?

rfratto replied:
This shouldn't happen because the new token isn't in the ring yet, but it's better to be safe here; I'll move the block.

owen-d commented on pkg/ingester/incremental_transfer.go:
```diff
@@ -0,0 +1,260 @@
+}
+
+// UnblockRanges manually removes blocks for the provided ranges.
+func (i *Ingester) UnblockRanges(ctx context.Context, in *client.UnblockRangesRequest) (*client.UnblockRangesResponse, error) {
```
should probably be `_ context.Context` since the arg is unused

owen-d commented on pkg/ingester/incremental_transfer.go:
```diff
@@ -0,0 +1,260 @@
+// configuration variable. This acts as a failsafe to prevent blocked ranges
+// sitting around forever if a joining ingester crashes, as writes will continue
+// to go to us and get rejected for as long as the blocked range exists.
+func (i *Ingester) BlockRanges(ranges []ring.TokenRange) {
```
One thing I'm worried about is transfers taking longer than `cfg.RangeBlockPeriod`. If it does, the the range would become unblocked due to this expiration and the ingester would become available for writes again.

What would happen to the new ingester receiving tokens? I think if it immediately retried, that would defeat the purpose of the `cfg.RangeBlockPeriod` as `BlockRanges` would be called again.

owen-d replied:
From @rfratto:

> 1. if a join fails, you'll still join the ring, but you won't have as much data as you should. it'll cause some spillover and desynced chunks.
> 2. if a leave fails, you flush to GCS. also causes some desynced chunks.
> 3. if a transfer is slow (or if a transferring node crashes), you risk more writes being rejected because the blocked range gets released later. we really need to set this value to the longest period we're comfortable tolerating rejections. the tradeoff here is the risk between rejecting data and having spillover.

These alleviated my concerns here.

rfratto commented on pkg/ingester/incremental_transfer.go:
```diff
@@ -0,0 +1,260 @@
+
+// TransferChunksSubset accepts chunks from a client and moves them into the local Ingester.
+func (i *Ingester) TransferChunksSubset(stream client.Ingester_TransferChunksSubsetServer) error {
+	i.userStatesMtx.Lock()
```
My understanding is that the lock is on the joining ingester so there wouldn't be any data to ingest anyway, right?

rfratto commented on pkg/ingester/ingester.go:
```diff
@@ -144,6 +176,9 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.BoolVar(&cfg.SpreadFlushes, "ingester.spread-flushes", false, "If true, spread series flushes across the whole period of MaxChunkAge")
 	f.IntVar(&cfg.ConcurrentFlushes, "ingester.concurrent-flushes", 50, "Number of concurrent goroutines flushing to dynamodb.")
+	f.BoolVar(&cfg.CheckOnCreate, "ingester.check-token-on-create", false, "Check that newly created streams fall within expected token ranges")
+	f.BoolVar(&cfg.CheckOnAppend, "ingester.check-token-on-append", false, "Check that existing streams appended to fall within expected token ranges")
```
I turn them all on for debugging. I'll change it to one single debug flag.

rfratto commented on pkg/ingester/ingester.go:
```diff
@@ -283,15 +351,15 @@ func (i *Ingester) Push(ctx old_ctx.Context, req *client.WriteRequest) (*client.
 	for _, ts := range req.Timeseries {
 		for _, s := range ts.Samples {
-			err := i.append(ctx, userID, ts.Labels, model.Time(s.TimestampMs), model.SampleValue(s.Value), req.Source)
+			err := i.append(ctx, userID, ts.Token, ts.Labels, model.Time(s.TimestampMs), model.SampleValue(s.Value), req.Source)
```
My understanding of this is one of two things will happen:

1. The tokens change, sharding to new ingesters. The old ingesters don't receive appends anymore and their data eventually gets flushed as underutilized chunks.
2. The tokens change, sharding to the same ingesters (coincidentally). The ingesters update the token in the existing memory series and log a warning.

rfratto commented on pkg/ingester/incremental_transfer.go:
```diff
@@ -0,0 +1,260 @@
+}
+
+// TransferChunksSubset accepts chunks from a client and moves them into the local Ingester.
+func (i *Ingester) TransferChunksSubset(stream client.Ingester_TransferChunksSubsetServer) error {
```
Yes, good idea.

rfratto commented on pkg/ring/ring.go:
```diff
@@ -211,15 +210,12 @@ func (r *Ring) Get(key uint32, op Operation, buf []IngesterDesc) (ReplicationSet
+		// is increased for the key. Dead ingesters will be filtered later by
+		// replication_strategy.go. Filtering later means that we can calculate
+		// a healthiness quorum.
+		if !ingester.IsHealthyState(op) {
```
Right, I'll change this back to its original logic but add in anther check for `ingester.Incremental`, allowing every non-PENDING state to be valid.

This check was also initially in `IsHealthyState` but got lost in a rebase. I'll have to add it back in on Monday.

bboreham commented:
Does this fix #467 and #775 ?

rfratto commented:
> Does this fix #467 and #775?
I'm not sure about #775, but it should fix #467.

rfratto commented on pkg/ingester/ingester.go:
```diff
@@ -654,3 +760,22 @@ func (i *Ingester) ReadinessHandler(w http.ResponseWriter, r *http.Request) {
 	}
 }
+
+func (i *Ingester) unexpectedStreamsHandler(tokens []uint32) {
```
would passing the unexpected series metric to the token checker be acceptable instead? I don't want to remove the logging and metric completely.

rfratto commented on pkg/ingester/incremental_transfer.go:
```diff
@@ -0,0 +1,263 @@
+)
+
+var (
+	blockedRanges = prometheus.NewGauge(prometheus.GaugeOpts{
```
I'm happy to do this, but my understanding is that the goal was to move away from promauto, wasn't it?

pracucci replied:
The goal is to move away from globally registered metrics. `promauto` is just syntactic sugar over what you're already doing. I haven't commented about passing the registry in input since this PR is already in a advanced state and didn't want to have you reconsider the design.

rfratto replied:
Ah, understood. I'll change it to `promauto`. Thanks!

rfratto commented:
Unfortunately, I think I need to close this PR and take another shot at this. I've rebased it to fix conflicts so many times now that I've lost confidence in its correctness, and I understand that its size has been a pain point for everyone all around. 

Rather than one giant PR, I'm going to start work on the first of a smaller set of PRs to eventually build up to this feature. Hopefully starting fresh will allow me to reduce the code complexity introduced here, although I still intend to be copying and pasting at least some of the existing code. 

Huge thank you to everyone who took time tackling this beast, and I hope the second attempt is a smoother experience for everyone :slightly_smiling_face:
